File: codingTest/programmers/DP/정수 삼각형.js
function solution(triangle) {
    var answer = 0;
    let dp = [[triangle[0][0]]]
    for (let i=1;i<triangle.length;i++) {
        dp[i]= []
        for (let j=0;j<triangle[i].length;j++) {
            if (j == 0) {
                dp[i][j] = dp[i-1][j]  + triangle[i][j]
            }
            else if (j == triangle[i].length-1) {
                dp[i][j] = dp[i-1][j-1] +  triangle[i][j] 
            }
            else {
                dp[i][j] = Math.max(dp[i-1][j-1],dp[i-1][j]) + triangle[i][j] 
            }
        }
    }
    return dp[triangle.length-1].reduce((a,b) => a>b?a:b);
    
}
// Math.max() 최댓값
// reduce()
// dp = [[]] 2차원 배열 
console.log(solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]));

File: codingTest/programmers/DP/정수 삼각형.py
# https://programmers.co.kr/learn/courses/30/lessons/43105
# 정수삼각형.py

def solution(triangle):
    
    length = len(triangle)
    triangle = list(reversed(triangle))
    dp = [[] for _ in range(length)]
    dp[0] = triangle[0]
    for i in range(1, length):
        for j in range(len(triangle[i])):
            dp[i].append(max(dp[i-1][j],dp[i-1][j+1]) + triangle[i][j])
    return dp[-1][0]